File: sonnet-scrambler/sonnetScrambler.js
var sonnetCounter = 154;

document.getElementById("scramblerButton").addEventListener("click", function() {
  document.getElementById("sonnetBody").innerHTML = writeSonnet();
  document.getElementById("buttonText").innerHTML = "Write Another Sonnet";
  sonnetCounter++;
  document.getElementById("sonnetNumber").innerHTML = "Sonnet " + sonnetCounter;
});

function writeSonnet() {
  var sonnet = "";
  var sonnetRhymes = [];
  var sonnetSources = [];
  //Select random couplets for the sonnet
  for (var i = 0; i < 7; i++) {
    var n = Math.floor(Math.random()*rhymeCouplets.length);
    sonnetRhymes.push(rhymeCouplets[n]);
    sonnetSources.push(Math.floor(n/7));
  }
  //Construct a string "sonnet" comprised of the sonnet in the correct format
  for (var i = 0; i < 5; i = i+2) {
    sonnet += sonnetRhymes[i][0] + "<br>" + sonnetRhymes[i+1][0] + "<br>" +
    sonnetRhymes[i][1] + "<br>" + sonnetRhymes[i+1][1]+ "<br>";
  }
  sonnet += "&nbsp;&nbsp;&nbsp;" + sonnetRhymes[6][0] + "<br>" + "&nbsp;&nbsp;&nbsp;" + sonnetRhymes[6][1];
  sonnetSources = sonnetSources.filter(function(elem, index, self) {
    return index == self.indexOf(elem);
  })
  //Add a sentence to "sonnet" listing the original source sonnets
  var sources = "<p class=\"sources\">Sonnet lines from";
  for (i = 0; i < sonnetSources.length; i++) {
    if (i < sonnetSources.length - 1) {
      sources += " " + sonnetSources[i] + ",";
    }
    else {
      sources += " & " + sonnetSources[i] + ".";
    }
  }
  return sonnet + sources;
}

File: sonnet-scrambler/sonnetScrambler.py
from random import randint
import json

# Make an array of rhyming couplets by:
# 1. Splitting the document into sonnets
# 2. Splitting each sonnet into lines (by "[p]" character)
# 3. Picking couplets according to strict sonnet structure (which is what Shakespeare used)
# Return the final array of couplets structured as:
# [[rhyme1a,rhyme1b],[rhyme2a, rhyme2b]...[rhyme1078a, rhyme 1078b]]

def make_rhyme_pairs_array(file):
    sonnet_file = open(file, 'r') #file = sonnets.txt or test.txt
    all_sonnets = sonnet_file.read()
    sonnet_file.close()
    sonnet_array = all_sonnets.split("\"\n\"")
    for i in range(len(sonnet_array)):
        sonnet_array[i]=sonnet_array[i].split("[p]")
    rhyme_pairs = []
    for sonnet in range(len(sonnet_array)):
        for line in range(len(sonnet_array[sonnet])):
            if line < 12 and line%4 < 2:
                rhyme_pairs.append([sonnet_array[sonnet][line],sonnet_array[sonnet][line+2]])
            elif line == 12:
                rhyme_pairs.append([sonnet_array[sonnet][line],sonnet_array[sonnet][line+1]])
    return rhyme_pairs

# Pick seven couplets at random and incorporate them into the sonnet structure.
def write_sonnet():
    rhyme_couplets = make_rhyme_pairs_array('sonnets.txt')
    sonnet = ""
    sonnet_rhymes = []
    sonnet_sources = []
    for i in range(7):
        n = randint(0,len(rhyme_couplets))
        sonnet_rhymes.append(rhyme_couplets[n])
        sonnet_sources.append(int(n/7))
    for i in [0, 2, 4]:
        sonnet += sonnet_rhymes[i][0] + sonnet_rhymes[i+1][0] + sonnet_rhymes[i][1] + sonnet_rhymes[i+1][1]
    sonnet += "  " + sonnet_rhymes[6][0] + "  " + sonnet_rhymes[6][1]
    sonnet_sources = list(set(sonnet_sources))
    sources = "\nSonnet lines from"
    for i in range(len(sonnet_sources)):
        if i < len(sonnet_sources)-1:
            sources += " %s," %sonnet_sources[i]
        else:
            sources += " & %s." %sonnet_sources[i]
    f = open('sonnet_output.txt', 'w')
    f.write(sonnet + sources)
    f.close()

write_sonnet()
